Improve terminology for Enhancers (#4593)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/Routing/Introduction.rst

For example it maps an URL like :samp:`https://example.org/news` to the News page. This process of
determining the page and/or action to execute for a specific URL is called "Routing".

The input of a Route is made up of several components; some components can also be split further
into sub-components.

Routing will also take care of beautifying URI parameters, for example converting
:samp:`https://example.org/profiles?user=magdalena` to :samp:`https://example.org/profiles/magdalena`.

Key Terminology
===============

> .. index:: Routing; Structure

Reviewer: ```suggestion
..  index:: Routing; Structure
```

| URL (no arguments)                                                                                                                                     |                |                |                        |
+--------------------------------------------------------------------------------------------------------------------------------------------------------+----------------+----------------+------------------------+
| URI (everything)                                                                                                                                                                                                  |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

..  hint::

    Please note that the following terminology is based on technical terms used in the TYPO3 Core,
    due to their Class/Object and Interface names.

> .. index:: Routing; Route

Reviewer: ```suggestion
..  index:: Routing; Route
```

Route
   The "speaking URL" as a whole (without the domain parts); for example `/en/about-us/our-team/john-doe/publications/index.xhtml`.
   This is also sometimes referred to as `Permalink`, some definitions also include the `URI arguments`
   for this term.

> .. index:: Routing; Site Language Prefix

Reviewer: ```suggestion
..  index:: Routing; Site Language Prefix
```

For example it maps an URL like :samp:`https://example.org/news` to the News page. This process of
determining the page and/or action to execute for a specific URL is called "Routing".

> The input of a Route is made up of several components; some components can also be split further

Reviewer 2: Why uppercase "Route"?

Reviewer 3: Fixed it

into sub-components.

Routing will also take care of beautifying URI parameters, for example converting
:samp:`https://example.org/profiles?user=magdalena` to :samp:`https://example.org/profiles/magdalena`.

..  _routing-terminology:

Key Terminology
===============

..  index:: Routing; Structure

Given a complex link (`URI`, `Uniform Resource Identificator`) like

    `https://subdomain.example.com:80/en/about-us/our-team/john-doe/publications/index.xhtml?utm_campaign=seo#start`

all of its components can be broken down to:

+----------+------------+----------+-----+------+----------------------+--------------------+----------------+----------------+-------+------------------+----------------+----------------+------------------------+
| https:// | subdomain. | example. | com | :80  | /en                  | /about-us/our-team | /john-doe      | /publications/ | index | .xhtml           | ?utm_campaign= | seo            | #start                 |
+==========+============+==========+=====+======+======================+====================+================+================+=======+==================+================+================+========================+

> | Protocol | Subdomain  | Domain   | TLD | Port | Site Language Prefix | Slug               | Enhanced Route                                             |                |                |                        |

Reviewer 2: Maybe

    abbr:`TLD (Top-Level Domain)`
    
? This explains the abbreviation (or write it out).

Additionally, why "Site Language Prefix" et al instead of "Site language prefix"?

Reviewer 3: Please make a follow up if desired.

+----------+------------+----------+-----+------+----------------------+--------------------+-----------------------------------------+------------------+----------------+----------------+------------------------+

> |          | Hostname                    |      |                      |                    | Route Enhancer                          | Route Decorator  | URI argument   | argument value | Location Hash / Anchor |

Author: You are totally right, the term "query string" is more appropriate

> |          | Hostname                    |      |                      |                    | Route Enhancer                          | Route Decorator  | Query string   | argument value | Location Hash / Anchor |

Reviewer 4: I understand what you want to show. But here "john-doe/publications/index.xhtml" will be processed but just ONE Route Enhancer. There is no Route Decorator. To realize these endings you need a Route Enhancer with decorator functionality

+-----------------------------------------------+--------------------------------------------------------------------------------------------------------+----------------+----------------+------------------------+
| URL (no arguments, unlike the URI)                                                                                                                                     |                |                |                        |
+--------------------------------------------------------------------------------------------------------------------------------------------------------+----------------+----------------+------------------------+
| URI (everything)                                                                                                                                                                                                  |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

..  hint::

    Please note that the following terminology is based on technical terms used in the TYPO3 Core,

>     due to their Class/Object and Interface names.

Reviewer 2: ```suggestion
    due to their class/object and interface names.
```

That are now common names, aren't they?

..  index:: Routing; Route

Route
   The "speaking URL" as a whole (without the domain parts); for example `/en/about-us/our-team/john-doe/publications/index.xhtml`.

>    This is also sometimes referred to as `Permalink`, some definitions also include the `URI arguments`

Reviewer 5: ```suggestion
   This is also sometimes referred to as `permalink`, some definitions also include the `URI arguments`
```

>    This is also sometimes referred to as `permalink`, some definitions also include the `Query string`

Reviewer 4: The query string is everything behind `?`. As the query path does not contain any `?` there is no query string in your example.

Author: In the table there is a query string; it is not part of a "speaking URL" by default, so the URL example before it doesn't contain it - it is explained later on so I think this shouldn't be a problem.

   for this term.

..  index:: Routing; Site Language Prefix

Site Language Prefix

>     A global site language prefix is not considered part of the slug, but rather as a "prefix" to the slug.

Reviewer 5: ```suggestion
    A global site language prefix (e.g. "/dk" or "/en-us") is not considered part of the slug, but rather a "prefix" to the slug.
```

> .. index:: Routing; Slug

Reviewer: ```suggestion
..  index:: Routing; Slug
```

 
Slug
    Unique name for a resource to use when creating URLs; for example the slug of the news detail page
    could be `/news/detail`, and
    the slug of a news record could be `2019-software-update`.

>     Within TYPO3, a slug is always a part (section) of the URL "path" - it does not contain scheme, host, HTTP verb, etc.

Reviewer 4: What is `verb`?

Author: GET | POST | PUT | DELETE ...

    * Slugs must be separated by one or more character like "/", "-", "_" and "&".
      Regular characters like letters should not be used as separators for better readability.

    ..  note::

        A slug of a record may contain slashes but this is not recommended:
        The risk of conflicts is higher when using slashes within slugs. For
        example, unrelated page hierarchies and records could have slugs
        forming the same URL path.

> .. index:: Routing; Enhancers

Reviewer: ```suggestion
..  index:: Routing; Enhancers
```

Enhancers
    Sections after a slug can be added ("enhanced") both by "Route Enhancers" and also 
    "(Route Enhancing) Decorators", see 
    :ref:`Advanced routing configuration <t3coreapi:routing-advanced-routing-configuration>`.

> .. index:: Routing; Page Type Suffix

Reviewer: ```suggestion
..  index:: Routing; Page Type Suffix
```

Page Type Suffix
    A Page Type Suffix indicates the type of a URL, usually ".html". It can also be left out completely.
    If set, it could control alternate variants of a URL, for example an RSS-Feed or a JSON representation.

    A Page Type Suffix can also be regarded as an Enhancer.

> .. index:: Routing; Enhanced Route

Reviewer: ```suggestion
..  index:: Routing; Enhanced Route
```

Enhanced Route
    The combination of multiple Enhancers (and the Page Type Suffix) can be referred to as the "Enhanced Route".

> .. index:: Routing; URI arguments

Reviewer: ```suggestion
..  index:: Routing; URI arguments
```

URI arguments
    The main distinction of `URL` (Uniform Resource Locator) and `URI` (Uniform Resource Identifier) is that
    the URI also include arguments/parameters and their values, beginning with a `?` and each argument
    separated by `&`, and the value separated from the argument name by `=`.

> .. index:: Routing; Location Hash

Reviewer: ```suggestion
..  index:: Routing; Location Hash
```

    A Page Type Suffix is regarded as an Enhancer, specifically a "(Route) Decorator".
    Other kinds of decorators could add additional parts to the route, but
    only after(!) the initial "Route Enhancer(s)".

.. index:: Routing; Enhanced Route

Enhanced Route
    The combination of multiple Enhancers (and the Page Type Suffix) can be referred to as the "Enhanced Route".

.. index:: Routing; URI arguments

URI arguments
    The main distinction of `URL` (Uniform Resource Locator) and `URI` (Uniform Resource Identifier) is that
    the URI also include arguments/parameters and their values, beginning with a `?` and each argument
    separated by `&`, and the value separated from the argument name by `=`.

.. index:: Routing; Location Hash
    The Location Hash is the part starting with `#` of a URI. Note that browsers requesting a resource
    never supply the hash to the Webserver, so any kind of Enhancer is not able to use this information
    to match any kind of routing. This can only be done by the Browser (for example via JavaScript),
    after the requested document has been rendered.

> .. _routing-terminology-symfony:

Reviewer: ```suggestion
..  _routing-terminology-symfony:
```

    example, unrelated page hierarchies and records could have slugs
    forming the same URL path.

..  index:: Routing; Enhancers

Enhancers
    Sections **after** a slug can be added ("enhancing" the route) both by "Route Enhancers" and also 
    "(Route Enhancing) Decorators", see 
    :ref:`Advanced routing configuration <t3coreapi:routing-advanced-routing-configuration>`.

..  index:: Routing; Page Type Suffix

Page Type Suffix
    A Page Type Suffix indicates the type of a URL, usually ".html". It can also be left out completely.

>     If set, it could control alternate variants of a URL, for example an RSS feed or a JSON representation.

Reviewer 4: ```suggestion
    If set, it could control alternate variants of a URL, for example a RSS feed or a JSON representation.
```

>     A Page Type Suffix is treated as an Enhancer, specifically a "(Route) Decorator".

Reviewer 4: An enhancer with decorator functionality

* Page Routing
* Route Enhancements and Aspects

Page Routing describes the process of resolving the concrete page (in earlier TYPO3 versions this were the `id` and `L` `$_GET` parameters,
now this uses the Site Language Prefix plus one or more slugs),
whereas Route Enhancements and Aspects take care of all additionally configured parameters (such as beautifying plugin parameters, handling `type` etc.).

Mathias Schreiber demonstrates this way of handling URLs
(Version 9.5, 28.09.2018).

> .. youtube:: dUz4B08XFes

Reviewer: ```suggestion
..  youtube:: dUz4B08XFes
```

* Page Routing
* Route Enhancements and Aspects

Page Routing describes the process of resolving the concrete page (in earlier TYPO3 versions this were the `id` and `L` `$_GET` parameters,
now this uses the Site Language Prefix plus one or more slugs),
whereas Route Enhancements and Aspects take care of all additionally configured parameters (such as beautifying plugin parameters, handling `type` etc.).

Mathias Schreiber demonstrates this way of handling URLs
(Version 9.5, 28.09.2018).

.. youtube:: dUz4B08XFes

> .. _routing-prerequisites:

Reviewer: ```suggestion
..  _routing-prerequisites:
```
